tools/wptrunner/wptrunner/browsers/edge.py

```diff
@@ -84,6 +85,12 @@ def start(self, **kwargs):
 
     def stop(self, force=False):
         self.server.stop(force=force)
```

Comment:
Things have already started to go wrong by this call to `self.server.stop`.

We call this function from: https://github.com/web-platform-tests/wpt/blob/2808e3adc6d5ee9bea8e0f1002a755d917ece2e7/tools/wptrunner/wptrunner/testrunner.py#L675

Note the line just before this: https://github.com/web-platform-tests/wpt/blob/2808e3adc6d5ee9bea8e0f1002a755d917ece2e7/tools/wptrunner/wptrunner/testrunner.py#L673

This `stop` message is sent *asynchronously* and goes through the [main testrunner loop](https://github.com/web-platform-tests/wpt/blob/2808e3adc6d5ee9bea8e0f1002a755d917ece2e7/tools/wptrunner/wptrunner/testrunner.py#L89), ending it, [making us return from running](https://github.com/web-platform-tests/wpt/blob/2808e3adc6d5ee9bea8e0f1002a755d917ece2e7/tools/wptrunner/wptrunner/testrunner.py#L151) and [exiting the context manager](https://github.com/web-platform-tests/wpt/blob/2808e3adc6d5ee9bea8e0f1002a755d917ece2e7/tools/wptrunner/wptrunner/testrunner.py#L68) where we follow our the `teardown` method including:

https://github.com/web-platform-tests/wpt/blob/2808e3adc6d5ee9bea8e0f1002a755d917ece2e7/tools/wptrunner/wptrunner/testrunner.py#L83

And that calls:

https://github.com/web-platform-tests/wpt/blob/2808e3adc6d5ee9bea8e0f1002a755d917ece2e7/tools/wptrunner/wptrunner/executors/executorwebdriver.py#L231

(Which just sends `DELETE /session/{session id}`)

Now, back on this process, note that this line (`self.server.stop`) calls [this](https://github.com/web-platform-tests/wpt/blob/2808e3adc6d5ee9bea8e0f1002a755d917ece2e7/tools/wptrunner/wptrunner/webdriver_server.py#L83), which on Windows calls `TerminateProcess` on the `MicrosoftWebDriver.exe`.

So we're racing `DELETE /session/{session id}` against outright terminating the process. This seems less than ideal.

As @mustjab said by email:

> It looks like the test rely on the driver to close sessions when it exits, but for Edge we don’t do that.

We shouldn't be relying on this. We should wait for the session being deleted (somehow!) before we terminate the WebDriver server. @jgraham do you have any suggestion as to how to implement this? We don't have access to the executor here, so we can't check if it's closed the session by running any idempotent command and seeing how if/how it fails.

Reply:
If calling `quit` in WebDriver ends up with the WebDriver process exiting then the `Browser` implementation for Edge can implement a `stop` function that initially just waits for the process, and only force-terminates it after a timeout. This is essentially what Firefox does. It's an assumption that you never call `stop` on the browser unless either you've already signalled for WebDriver to stop the process, or `force` is set to `True`, which is not ideal but works OK in practice.